### PR TITLE
Focus correctness improvements

### LINKED ISSUE #26900: Removing a focused element from the tree doesn't reset focus
```
<body>
  <style>div:focus { color: red; }</style>
  <div tabindex=0 id="a" style="background-color: green; width: 100px; height: 100px;"><button onclick="this.parentNode.remove()">remove</button><p>hello</p></div>
<script>
let div = document.querySelector('#a');
div.focus();
div.onkeydown = (e) => { console.log('div down ' + e.key); };
div.onkeyup = (e) => { console.log('div up ' + e.key); };
document.body.onkeydown = function(e) {
  console.log("body down " + e.key);
};
document.body.onkeyup = function(e) {
  console.log("body up " + e.key);
};
</script>
```
In Firefox, when I press `a`, then click the `remove` button and press `a` again, I see:
```
div down a
body down a
div up a
body up a
body down a
body up a
```
Which indicates that the focus was reset such that the div that was removed from the tree is no longer focused.
In Servo, doing the same gives me:
```
div down a
body down a
div up a
body up a
div down a
div up a
```
Which indicates that we don't reset the focus when the div is removed from the tree.

### PR DESCRIPTION
**Pull request by jdm**
These changes improve the behaviour of focus in Hubs rooms, and are expected to improve web compat around other dynamic pages that receive keyboard events as well.

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #26901 and fix #26900.
- [x] There are tests for these changes

### CODE REVIEW

**Comment by CYBAI** [comment]
tidy is complaining about the spec urls 👀 

```
Checking files for tidiness...
./components/script/dom/element.rs:1871: link to WHATWG may break in the future, use this format instead: https://html.spec.whatwg.org/multipage/#focus-fixup-rule
./components/script/dom/document.rs:1028: link to WHATWG may break in the future, use this format instead: https://html.spec.whatwg.org/multipage/#focus-fixup-rule
```

<details>

<summary>Open to see WPT failures</summary>

```
  ▶ Unexpected subtest result in /html/interaction/focus/the-autofocus-attribute/focusable-area-in-top-document.html:
  │ FAIL [expected PASS] If topDocument's focused area is not topDocument, autofocus is not processed.
  │   → assert_equals: Prereq: IFRAME should be focused expected Element node <iframe srcdoc="<input><script>document.querySelector('in... but got Element node <body><iframe srcdoc="<input><script>document.querySelect...
  │ 
  │ @http://web-platform.test:8000/html/interaction/focus/the-autofocus-attribute/focusable-area-in-top-document.html:14:16
  │ async*Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  │ promise_test/tests.promise_tests</<@http://web-platform.test:8000/resources/testharness.js:592:36
  └ promise_test/tests.promise_tests<@http://web-platform.test:8000/resources/testharness.js:591:20
  ▶ TIMEOUT [expected OK] /html/browsers/the-window-object/focus.window.html
  │ 
  │ error: XDG_RUNTIME_DIR not set in the environment.
  └ libEGL warning: No hardware driver found, falling back to software rendering
  ▶ Unexpected subtest result in /html/browsers/the-window-object/focus.window.html:
  │ TIMEOUT [expected FAIL] focus
  └   → Test timed out
  ▶ Unexpected subtest result in /html/interaction/focus/focus-file-input.html:
  │ FAIL [expected PASS] HTML Test: file input can be programatically focused before layout
  │   → assert_equals: expected Element node <input type="file"></input> but got Element node <body><input type="file">
  │   → <script>
  │   → test(function() {
  │   →   le...
  │ 
  │ @http://web-platform.test:8000/html/interaction/focus/focus-file-input.html:16:16
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  │ test@http://web-platform.test:8000/resources/testharness.js:535:30
  └ @http://web-platform.test:8000/html/interaction/focus/focus-file-input.html:12:5
  ▶ Unexpected subtest result in /html/interaction/focus/focus-management/focus-event-targets-simple.html:
  │ FAIL [expected PASS] Focus events fire at correct targets in correct order in simple case
  │   → assert_array_equals: Exactly 1 focus event should fire and its target should be the input lengths differ, expected array [Element node <input type="text" id="a"></input>] length 1, got [] length 0
  │ 
  │ @http://web-platform.test:8000/html/interaction/focus/focus-management/focus-event-targets-simple.html:29:22
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  │ test@http://web-platform.test:8000/resources/testharness.js:535:30
  └ @http://web-platform.test:8000/html/interaction/focus/focus-management/focus-event-targets-simple.html:28:5
  ▶ TIMEOUT [expected OK] /html/interaction/focus/focus-management/focus-events.html
  │ 
  │ error: XDG_RUNTIME_DIR not set in the environment.
  └ libEGL warning: No hardware driver found, falling back to software rendering
  ▶ Unexpected subtest result in /html/interaction/focus/focus-management/focus-events.html:
  └ NOTRUN [expected PASS] focusing on a focusable element fires a focus event at the element
  ▶ Unexpected subtest result in /html/interaction/focus/focus-management/focus-events.html:
  └ NOTRUN [expected PASS] focusing on a focusable element fires a blur event at the previous focussed element
  ▶ TIMEOUT [expected OK] /html/interaction/focus/composed.window.html
  │ 
  │ error: XDG_RUNTIME_DIR not set in the environment.
  └ libEGL warning: No hardware driver found, falling back to software rendering
  ▶ Unexpected subtest result in /html/interaction/focus/composed.window.html:
  │ TIMEOUT [expected FAIL] Focus events are composed
  └   → Test timed out
  ▶ Unexpected subtest result in /html/interaction/focus/tabindex-focus-flag.html:
  └ PASS [expected FAIL] a should not be focusable by default.
  ▶ Unexpected subtest result in /html/interaction/focus/tabindex-focus-flag.html:
  └ PASS [expected FAIL] input[type="hidden"] should not be focusable by default.
  ▶ Unexpected subtest result in /html/interaction/focus/tabindex-focus-flag.html:
  │ FAIL [expected PASS] input[type="button"] should be focusable by default.
  │   → assert_equals: expected Element node <input type="button"></input> but got Element node <body>
  │   → <div id="default-samples">
  │   → <a></a>
  │   → <a href=""></a>...
  │ 
  │ runTests/<@http://web-platform.test:8000/html/interaction/focus/tabindex-focus-flag.html:49:22
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  │ test@http://web-platform.test:8000/resources/testharness.js:535:30
  └ runTests@http://web-platform.test:8000/html/interaction/focus/tabindex-focus-flag.html:45:9
  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/execution-timing/031.html:
  │ FAIL [expected PASS]  scheduler: focus and blur events
  │   → assert_array_equals: lengths differ, expected array ["inline script #1", "focus event", "blur event", "focus el 2 event", "end script #1", "inline script #2"] length 6, got ["inline script #1", "end script #1", "inline script #2"] length 3
  │ 
  │ test@http://web-platform.test:8000/html/semantics/scripting-1/the-script-element/execution-timing/031.html:24:32
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:2002:35
  ▶ Unexpected subtest result in /html/semantics/forms/the-input-element/time-focus-dynamic-value-change.html:
  │ FAIL [expected PASS] Time input handles focus correctly when value changes
  │   → assert_true: expected true got false
  │ 
  │ window.onload<@http://web-platform.test:8000/html/semantics/forms/the-input-element/time-focus-dynamic-value-change.html:18:14
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:2018:32
  ▶ Unexpected subtest result in /html/semantics/selectors/pseudo-classes/focus.html:
  │ FAIL [expected PASS] input1 has the focus
  │   → assert_array_equals: lengths differ, expected array [Element node <input id="input1"></input>] length 1, got [] length 0
  │ 
  │ testSelectorIdsMatch/<@http://web-platform.test:8000/html/semantics/selectors/pseudo-classes/utils.js:12:24
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  │ test@http://web-platform.test:8000/resources/testharness.js:535:30
  │ testSelectorIdsMatch@http://web-platform.test:8000/html/semantics/selectors/pseudo-classes/utils.js:10:7
  │ run_test@http://web-platform.test:8000/html/semantics/selectors/pseudo-classes/focus.html:34:27
  └ onload@http://web-platform.test:8000/html/semantics/selectors/pseudo-classes/focus.html:26:9
  ▶ FAIL [expected PASS] /_mozilla/css/input_insertion_point_empty_a.html
  │   → /_mozilla/css/input_insertion_point_empty_a.html ['c341dbce9b98cfd7b7194034e55fcc6838cd55d3']
  └   → /_mozilla/css/input_insertion_point_empty_ref.html ['c341dbce9b98cfd7b7194034e55fcc6838cd55d3']
  ▶ FAIL [expected PASS] /_mozilla/css/input_selection_a.html
  │   → /_mozilla/css/input_selection_a.html ['54a9df64f1476dd12020019d7cf22ac34d727bc0']
  │   → /_mozilla/css/input_selection_ref.html ['caba1bd18437689eecf901bb4f9a0a9fdb4b8ae6']
  └   → Screenshot is solid color 0xFFFFFF for /_mozilla/css/input_selection_a.html
  ▶ FAIL [expected PASS] /_mozilla/css/input_selection_incremental_a.html
  │   → /_mozilla/css/input_selection_incremental_a.html ['54a9df64f1476dd12020019d7cf22ac34d727bc0']
  │   → /_mozilla/css/input_selection_incremental_ref.html ['e5c995e16a7c3c067520757b978ca12e5f181dca']
  └   → Screenshot is solid color 0xFFFFFF for /_mozilla/css/input_selection_incremental_a.html
  ▶ FAIL [expected PASS] /_mozilla/css/submit_focus_a.html
  │   → /_mozilla/css/submit_focus_a.html ['a1d3a41917e1205b302b8eeb08b6958b79ee9365']
  └   → /_mozilla/css/submit_focus_b.html ['17d293fd5157a401f99ec2bbd5c81224dc3c8c0c']
  ▶ Unexpected subtest result in /html/interaction/focus/document-level-focus-apis/document-level-apis.html:
  │ FAIL [expected PASS] The element must be the active element if it is focused
  │   → assert_equals: The active element should be the input element. expected Element node <input id="test"></input> but got Element node <body><div id="log"></div>
  │   → <input id="test">
  │   → <script>
  │   → 
  │   → te...
  │ 
  │ @http://web-platform.test:8000/html/interaction/focus/document-level-focus-apis/document-level-apis.html:18:16
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  │ test@http://web-platform.test:8000/resources/testharness.js:535:30
  └ @http://web-platform.test:8000/html/interaction/focus/document-level-focus-apis/document-level-apis.html:16:5
  ▶ Unexpected subtest result in /html/interaction/focus/processing-model/focus-fixup-rule-one-no-dialogs.html:
  └ PASS [expected FAIL] Disabling <fieldset> affects its descendants
  ▶ Unexpected subtest result in /html/interaction/focus/processing-model/focus-fixup-rule-one-no-dialogs.html:
  │ FAIL [expected PASS] Changing the first legend element in disabled <fieldset>
  │   → assert_not_equals: After changing a legend element, the button must no longer be focused got disallowed value Element node <button id="button5">Button 5</button>
  │ 
  │ @http://web-platform.test:8000/html/interaction/focus/processing-model/focus-fixup-rule-one-no-dialogs.html:82:20
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  │ test@http://web-platform.test:8000/resources/testharness.js:535:30
  └ @http://web-platform.test:8000/html/interaction/focus/processing-model/focus-fixup-rule-one-no-dialogs.html:74:5
  ▶ TIMEOUT [expected OK] /_mozilla/mozilla/FocusEvent.html
  │ 
  │ error: XDG_RUNTIME_DIR not set in the environment.
  └ libEGL warning: No hardware driver found, falling back to software rendering
  ▶ Unexpected subtest result in /_mozilla/mozilla/FocusEvent.html:
  │ TIMEOUT [expected PASS] FocusEvent
  └   → Test timed out
  ▶ TIMEOUT [expected OK] /_mozilla/mozilla/double_focus.html
  │ 
  │ error: XDG_RUNTIME_DIR not set in the environment.
  └ libEGL warning: No hardware driver found, falling back to software rendering
  ▶ Unexpected subtest result in /_mozilla/mozilla/double_focus.html:
  │ TIMEOUT [expected PASS] Double focus/blur events
  └   → Test timed out
  ▶ Unexpected subtest result in /_mozilla/mozilla/document_activeElement.html:
  │ FAIL [expected PASS] document_activeElement
  │   → assert_equals: test_2.2, document.activeElement expected Element node <input id="foo" type="text"></input> but got Element node <body>
  │   →         <input id="foo" type="text">
  │   →         <scri...
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/document_activeElement.html:17:26
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  │ test@http://web-platform.test:8000/resources/testharness.js:535:30
  └ @http://web-platform.test:8000/_mozilla/mozilla/document_activeElement.html:11:15
  ▶ Unexpected subtest result in /html/interaction/focus/processing-model/preventScroll-textarea.html:
  │ FAIL [expected PASS] preventScroll: true on a textarea element
  │   → assert_equals: INPUT: Should have been focused expected Element node <input value="EFGH"></input> but got Element node <body><div style="height: 200vh"></div>
  │   → <textarea>ABCD</t...
  │ 
  │ @http://web-platform.test:8000/html/interaction/focus/processing-model/preventScroll-textarea.html:23:18
  │ async*Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  │ promise_test/tests.promise_tests</<@http://web-platform.test:8000/resources/testharness.js:592:36
  └ promise_test/tests.promise_tests<@http://web-platform.test:8000/resources/testharness.js:591:20
  ▶ Unexpected subtest result in /_mozilla/mozilla/focus_blur.html:
  │ FAIL [expected PASS] focus_blur
  │   → assert_equals: expected Element node <input id="a"></input> but got Element node <body>
  │   →   <input id="a">
  │   →   <input id="b">
  │   →   <script>
  │   →   tes...
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/focus_blur.html:18:18
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  │ test@http://web-platform.test:8000/resources/testharness.js:535:30
  └ @http://web-platform.test:8000/_mozilla/mozilla/focus_blur.html:12:7
```

</details>

**Comment by CYBAI** [comment]
the failure is #25476

File: components/script/dom/element.rs [hunk 1]
```diff
@@ -1832,6 +1832,67 @@ impl Element {
...
+            )) |
+            NodeTypeId::Element(ElementTypeId::HTMLElement(
+                HTMLElementTypeId::HTMLSelectElement,
+            )) |
+            NodeTypeId::Element(ElementTypeId::HTMLElement(
+                HTMLElementTypeId::HTMLIFrameElement,
+            )) |
+            NodeTypeId::Element(ElementTypeId::HTMLElement(
+                HTMLElementTypeId::HTMLTextAreaElement,
+            )) => true,
```
Comment (nox) [inline components/script/dom/element.rs:1868]: Just a random thought: would be nice to generate associated constants on `NodeTypeId` to do stuff such as `NodeTypeID::TEXT_AREA`.

File: components/script/dom/document.rs [hunk 2]
```diff
@@ -1011,21 +1021,50 @@ impl Document {
...
-            self.possibly_focused.set(Some(elem))
+    pub(crate) fn request_focus(&self, elem: Option<&Element>, focus_type: FocusType) {
+        let implicit_transaction = matches!(
+            *self.focus_transaction.borrow(),
+            FocusTransaction::NotInTransaction
+        );
+        if implicit_transaction {
+            self.begin_focus_transaction();
+        }
+        if elem.map_or(true, |e| e.is_focusable_area()) {
```
Comment (nox) [inline components/script/dom/document.rs:1048]: I'm not sure I understand what it means to focus a `None` element.

**Comment by nox** [comment]
LGTM but there was a try failure.

File: components/script/dom/document.rs (hunk 2 above)
Comment (jdm) [reply components/script/dom/document.rs:1048]: De-focus the currently focused element.

File: components/script/dom/element.rs (hunk 1 above)
Comment (jdm) [reply components/script/dom/element.rs:1868]: Filed #26925.

**Pull request merged**